src/lib/device/journal_vdev.hpp

```diff
@@ -19,250 +19,367 @@
+         * @return : On success, the number of bytes read is returned (zero indicates end of file), and the cursor is
+         * advanced by this number. it is not an error if this number is smaller than the number requested, because it
+         * can be end of chunk, since read won't across chunk.
+         */
+        size_t sync_next_read(uint8_t* buf, size_t count_in);
+
+        /**
+         * @brief : reads up to count bytes at offset into the buffer starting at buf.
+         * The curosr is not updated.
+         *
+         * @param buf : the buffer that points to the read out data.
+         * @param count : size of buffer
+         * @param offset : the start offset to do read
+         *
+         * @return : return the error code of the read
+         */
+        std::error_code sync_pread(uint8_t* buf, size_t count_in, off_t offset);
+
+        /**
+         * @brief : read at offset and save output to iov.
+         * We don't have a use case for external caller of preadv now, meaning iov will always have only 1 element;
+         * if the len is acrossing chunk boundary,
+         * we only do read on one chunk and return the num of bytes read on this chunk;
+         *
+         * @param iov : the iovect to store the read out data
+         * @param iovcnt : size of iovev
+         * @param offset : the start offset to read
+         *
+         * @return : return the error code of the read
+         */
+        std::error_code sync_preadv(iovec* iov, int iovcnt, off_t offset);
+
+        /**
+         * @brief : repositions the cusor of the device to the argument offset
+         * according to the directive whence as follows:
+         * SEEK_SET
+         *     The curosr is set to offset bytes.
+         * SEEK_CUR
+         *     The cursor is set to its current location plus offset bytes.
+         * SEEK_END
+         *     Not supported yet. No use case for now.
+         *
+         * @param offset : the logical offset
+         * @param whence : see above
+         *
+         * @return :  Upon successful completion, lseek() returns the resulting offset
+         * location as measured in bytes from the beginning of the file.  On
+         * error, the value (off_t) -1 is returned
+         */
+        off_t lseek(off_t offset, int whence = SEEK_SET);
+
+        /**
+         * @brief : this API can be replaced by lseek(0, SEEK_CUR);
+         *
+         * @return : current curosr offset
+         */
+        off_t seeked_pos() const { return m_seek_cursor; }
+
+        /**
+         * @brief :- it returns the vdev offset after nbytes from start offset
+         */
+        off_t dev_offset(off_t nbytes) const;
+
+        /**
+         * @brief : get the start logical offset where data starts;
+         *
+         * @return : the start logical offset where data starts;
+         */
+        off_t data_start_offset() const { return m_data_start_offset; }
+
+        off_t end_offset() const { return m_end_offset; }
```

Comment:
can you also provide comment of this API, especially the difference between tail_offset, so that who are not familiar with this code can be easier to understand.

Reply:
Its added near the class variables. journal_vdev.hpp:50. If its not clear, please let me know.